lib/common/bits.h

```diff
@@ -0,0 +1,244 @@
+                                                     1, 3, 4, 4, 2, 5, 6, 7,
+                                                     7, 0, 1, 2, 3, 3, 4, 6,
+                                                     2, 6, 5, 5, 3, 4, 5, 6,
+                                                     7, 1, 2, 4, 6, 4, 4, 5,
+                                                     7, 2, 6, 5, 7, 6, 7, 7 };
+            return DeBruijnBytePos[((U64)((val & -(long long)val) * 0x0218A392CDABBD3FULL)) >> 58];
+        } else { /* 32 bits */
+            static const int DeBruijnBytePos[32] = { 0, 0, 3, 0, 3, 1, 3, 0,
+                                                     3, 2, 2, 1, 3, 2, 0, 1,
+                                                     3, 3, 1, 2, 2, 2, 2, 0,
+                                                     3, 1, 2, 0, 1, 0, 1, 1 };
+            return DeBruijnBytePos[((U32)((val & -(S32)val) * 0x077CB531U)) >> 27];
+        }
+    } else {  /* Big Endian CPU */
+        unsigned r;
+        if (MEM_64bits()) {
+            const unsigned n32 = sizeof(size_t)*4;   /* calculate this way due to compiler complaining in 32-bits mode */
+            if (!(val>>n32)) { r=4; } else { r=0; val>>=n32; }
+            if (!(val>>16)) { r+=2; val>>=8; } else { val>>=24; }
+            r += (!val);
+            return r;
+        } else { /* 32 bits */
+            if (!(val>>16)) { r=2; val>>=8; } else { r=0; val>>=24; }
+            r += (!val);
+            return r;
+        }
+    }
+}
+
+MEM_STATIC unsigned ZSTD_NbCommonBytes(size_t val)
+{
+    if (MEM_isLittleEndian()) {
+        if (MEM_64bits()) {
+#           if defined(_MSC_VER) && defined(_WIN64)
+#               if STATIC_BMI2
+                    return _tzcnt_u64(val) >> 3;
+#               else
+                    if (val != 0) {
+                        unsigned long r;
+                        _BitScanForward64(&r, (U64)val);
+                        return (unsigned)(r >> 3);
+                    } else {
+                        /* Should not reach this code path */
+                        __assume(0);
+                    }
+#               endif
+#           elif defined(__GNUC__) && (__GNUC__ >= 4)
+                return (unsigned)(__builtin_ctzll((U64)val) >> 3);
+#           else
+                return ZSTD_NbCommonBytes_fallback(val);
+#           endif
+        } else { /* 32 bits */
+#           if defined(_MSC_VER)
+                if (val != 0) {
+                    unsigned long r;
+                    _BitScanForward(&r, (U32)val);
+                    return (unsigned)(r >> 3);
+                } else {
+                    /* Should not reach this code path */
+                    __assume(0);
+                }
+#           elif defined(__GNUC__) && (__GNUC__ >= 3)
+                return (unsigned)(__builtin_ctz((U32)val) >> 3);
+#           else
+                return ZSTD_NbCommonBytes_fallback(val);
+#           endif
+        }
```

Comment:
E.g. the LE branch would be this.

```suggestion
        if (MEM_64bits()) {
            return ZSTD_countTrailingZeros64((U64)val) >> 3;
        } else { /* 32 bits */
            return ZSTD_countTrailingZeros32((U32)val) >> 3;
        }
```

And the BE branch would be the same but with `ZSTD_countLeadingZeros*()`.